### ISSUE TITLE
NVIDIA-01D Work Order — CUDA midpoint synthesis proof

### ISSUE THREAD

**Issue by piuteu** [piuteu/OBS-FrameGeneration-NVIDIA#10]
# NVIDIA-01D Work Order — CUDA midpoint synthesis proof

## Objective

Implement the fourth executable checkpoint of the NVIDIA-native OBS frame-generation project.

NVIDIA-01D proves only this:

> For each valid accepted A/B pair whose raw NVOFA forward/backward flow is complete, a fixed `t=0.5` midpoint frame can be synthesized entirely on the GPU, written to a persistent D3D11/OBS-visible output with correct pair ownership, and exposed in a deterministic real/generated presentation sequence.

This is the **first generated-frame foundation checkpoint**. The goal is to make the end-to-end midpoint path work. Do not turn this phase into final artifact removal, occlusion research, or performance tuning.

---

# 0. Read before coding

Read the current `main` versions of:

- `ACTIVE_DECISIONS.md`
- `PROJECT_STATUS.md`
- `TODO.md`
- `docs/NVIDIA01_IMPLEMENTATION_SPEC.md`
- `docs/NVIDIA01C_NVOFA_CUDA_PROOF.md`
- `docs/NVIDIA01C_PRIMARY_SOURCE_NOTES.md`
- GitHub Issue #9 — `NVIDIA-01D Reference Cross-Check — midpoint flow warp / synthesis`

Preserve the project rules already established:

- new NVIDIA-native design; no default old-build reuse;
- raw NVOFA CUDA path remains the flow source;
- no NvOFFRUC implementation;
- no Generic/DirectML/CPU fallback;
- no backend cascade;
- no automatic preset/grid/quality downgrade;
- no generated-frame skip or real-frame substitution to conceal a performance miss;
- no full-frame CPU readback/upload in the normal path.

### Primary-source / research rule

Before finalizing flow-grid reconstruction or CUDA warp coordinates, inspect the actual local NVIDIA Optical Flow SDK 5.0.7 package again for any official helper/sample code relevant to flow upsampling / grid interpretation.

Do **not** infer the precise spatial phase of the 4x4 vector grid merely from the phrase "one vector per 4x4 block" if the actual SDK package gives a more exact convention.

NVIDIA's official DALI optical-flow documentation is authoritative that full-resolution use of a 4x4 output flow upsamples the flow field **without multiplying the vector quantities by 4**.

Issue #9 contains the peer-reviewed midpoint-flow formula and backward-warp convention to cross-check before implementation.

---

# 1. Existing 01C foundation that must be reused

01C already provides:

- accepted A/B sequence and pair IDs;
- four persistent NVOF-owned ABGR8 input CUARRAY slots;
- persistent forward/backward `SHORT2` / S10.5 flow output sets in CUDA device memory;
- optional UINT8 cost on supported hardware;
- asynchronous NVOFA execution;
- explicit input/output-set ownership;
- output-stream readiness events;
- persistent CUDA/D3D11 interop;
- no full-frame CPU staging.

Do not recompute optical flow in 01D.

Do not copy A/B back from the D3D11 capture path again solely for synthesis if the already-owned NVOF input CUARRAY slots can safely serve as the synthesis source images.

The ideal ownership chain is:

```text
accepted A/B input slots
  -> NVOFA Execute
  -> FWD/BWD flow ready
  -> CUDA midpoint synthesis uses the same immutable A/B input slots
  -> generated output ready
  -> only then release pair input/output ownership for reuse
```

If the official API/resource semantics prevent direct synthesis reads from the existing NVOF input CUARRAYs, document the concrete reason and use the smallest GPU-only alternative. Do not introduce CPU staging.

---

# 2. Flow decode and full-resolution reconstruction

NVOFA flow values are signed S10.5 source-pixel displacement.

Required baseline:

```text
float flow_pixels = int16_value / 32.0f
```

For the 4x4 NVOFA output grid:

- reconstruct/sample the coarse vector field at full-resolution output coordinates;
- use bilinear flow interpolation unless the actual SDK 5.0.7 helper/sample gives a different authoritative method;
- preserve source-pixel displacement magnitude;
- **do not multiply the vector by 4 after upsampling**;
- respect the real returned output stride; never assume tightly packed rows.

You may reconstruct a persistent full-resolution flow buffer or sample the coarse flow field directly inside the midpoint kernel. Prefer the smaller/simple path unless measurement proves a separate dense-flow pass is necessary.

Do not add a custom flow pyramid.

---

# 3. Fixed midpoint math — t = 0.5 only

Use the non-learned bidirectional intermediate-flow approximation documented in Issue #9 / Super SloMo as the baseline.

Let:

- `F_AB` = NVOFA forward A -> B
- `F_BA` = NVOFA backward B -> A

For `t = 0.5`:

```text
F_mid_to_A = -0.25 * F_AB + 0.25 * F_BA
F_mid_to_B =  0.25 * F_AB - 0.25 * F_BA
```

Then backward-warp the input images:

```text
sample_A_position = midpoint_pixel + F_mid_to_A
sample_B_position = midpoint_pixel + F_mid_to_B
```

Use bilinear source-image sampling.

Initial color fusion when both samples are valid:

```text
midpoint = 0.5 * warped_A + 0.5 * warped_B
```

This is deliberately a simple foundation baseline.

Do not add learned flow refinement, learned visibility, softmax splatting, depth weighting, or a complex occlusion model.

---

# 4. Border / invalid sample handling

No kernel may perform an out-of-bounds image read.

Choose one small deterministic border policy supported cleanly by the selected CUDA sampling path (for example clamp-to-edge or explicit in-bounds weighting) and document it.

This is local sampling behavior, not a whole-frame fallback.

Do not build hole-filling/dilation/inpainting machinery in 01D.

Expected border/disocclusion artifacts are not a merge blocker if the midpoint transport/math is correct.

---

# 5. Confidence, consistency and NVOFA cost — keep out of the critical path

01D must not be blocked on final confidence/occlusion quality.

For this first midpoint baseline:

- NVOFA cost may remain available but synthesis must not require it;
- forward/backward consistency may be logged or prototyped only if it is needed to diagnose a correctness failure;
- do not stack cost + consistency + morphology + edge masks preemptively.

If the simple midpoint works but motion boundaries ghost, classify that as a later quality artifact and move on.

The TODO items for consistency/cost are **not 01D merge blockers** unless a concrete basic-correctness problem requires the smallest such check.

---

# 6. Native CUDA synthesis path

Implement the smallest explicit CUDA synthesis component appropriate for this project.

Preferred characteristics:

- native CUDA kernel(s), not TensorRT/PyTorch/ncnn;
- persistent resources sized on initialization/resize;
- no per-pair full-resolution allocation;
- no per-frame CUDA registration;
- no full-frame host staging;
- no unconditional `cudaDeviceSynchronize()`;
- no unconditional hot-path `cudaStreamSynchronize()`.

The kernel must consume:

- A and B GPU-resident source images;
- forward/backward flow pointers + actual strides;
- image dimensions / flow dimensions;
- fixed `t=0.5`.

Output must be full-resolution 8-bit color suitable for the existing OBS/D3D11 path.

If a persistent CUDA-owned RGBA output buffer followed by one measured D2D copy into the mapped D3D11 output is the simplest safe design, that copy is allowed. Do not force direct surface writes merely to claim literal zero-copy.

---

# 7. Synchronization and ownership

The output of `NvOFExecute` is asynchronous.

Synthesis must be correctly ordered after NVOFA output completion without a device-wide wait.

Preferred model:

```text
NVOFA output stream ordering / event
  -> midpoint CUDA work
  -> generated-output D2D/writeback
  -> generated-ready event / D3D11 unmap boundary
```

If synthesis uses a separate CUDA stream, bridge it with a CUDA event/device-side wait. Do not block the CPU merely to wait for flow.

### Important ownership change from 01C

01C currently releases A/B input-slot ownership when its flow output-ready generation completes.

01D must extend ownership so source A/B slots and the associated flow output set cannot be reused until **midpoint synthesis no longer reads them**.

Do not copy the whole flow/image to another buffer merely to avoid fixing ownership correctly.

Maintain explicit:

- pair ID;
- A sequence / B sequence;
- NVOF input slots;
- NVOF flow output set;
- generated output slot;
- generation/ready state.

A generated texture for pair N must never be presented as belonging to pair N+1.

---

# 8. Persistent generated output

Use a small fixed persistent generated-output pool compatible with the existing D3D11/CUDA interop. Two generated slots / ping-pong is the expected starting point unless the existing two 01B output slots can be cleanly repurposed without breaking ownership.

Requirements:

- no per-pair D3D11 texture allocation;
- CUDA-written generated frame becomes OBS-visible only after CUDA ownership ends;
- resize/format/device reset recreates the required output resources explicitly;
- no adaptive output-buffer count.

Do not preserve the 01B pass-through output architecture if it obstructs the actual generated-frame path. 01B was a proof; 01D may evolve that proof into the real NVIDIA output path while preserving its validated interop principles.

---

# 9. Minimal deterministic presentation state

01D should expose enough presentation state to prove the intended fixed-2x ordering:

```text
A(real) -> M(A,B,0.5) -> B(real) -> M(B,C,0.5) -> C(real) ...
```

This checkpoint does **not** need the full sustained 60 FPS deadline validation; that is NVIDIA-01E.

But 01D must prove that:

- a generated midpoint is associated with the correct pair;
- real/generated ordering is deterministic;
- no stale midpoint is reused for another pair;
- first-frame/history-reset has an explicit state;
- resize/reset discards stale generated state.

Do not add an adaptive scheduler, frame skipping, or real-frame substitution.

If complete live 30->60 cadence integration materially complicates the CUDA synthesis proof, implement the smallest deterministic delayed presentation state needed for a controlled OBS smoke and leave sustained timing/deadline proof to 01E.

---

# 10. Validation — foundation correctness only

## A. Static pair

A and B identical.

Expected:

- generated midpoint exists;
- sampled/checked midpoint pixels match the static input within the expected 8-bit/bilinear tolerance;
- no non-zero-motion requirement is invented.

## B. Known translation midpoint

Reuse/improve the deterministic textured feature field from 01C.

Use a known endpoint translation such as:

```text
A -> B = +24 px X, 0 px Y
```

Expected midpoint at `t=0.5` is approximately:

```text
+12 px X, 0 px Y
```

Validate several deterministic interior sample locations or a small GPU-reduced error statistic. Do not rely on one ambiguous textureless pixel.

The test should prove:

- S10.5 decode is correct;
- no erroneous x4 vector scaling exists;
- midpoint-flow signs are correct;
- backward-warp sample convention is correct;
- channel order/output format is correct.

Use broad correctness tolerance. This is not a PSNR/quality benchmark.

## C. Pair ownership

For at least two consecutive valid pairs:

- one generated midpoint per pair;
- midpoint pair ID matches its A/B sequences;
- no input/output slot reuse while synthesis is in flight.

## D. OBS-visible generated output

Render an actual generated midpoint through the persistent D3D11 OBS output path and verify a deterministic sample/pattern expected from the midpoint, not merely the old 01B CUDA test pattern/pass-through.

## E. Real/generated sequence

Exercise a deterministic controlled sequence and prove the logical order:

```text
real A -> midpoint AB -> real B -> midpoint BC
```

A short controlled smoke is sufficient. Sustained 60Hz contract belongs to 01E.

## F. Resize lifecycle

Exercise `1920x1080 -> 1280x720` or equivalent:

- stop/release in-flight synthesis safely at the defined lifecycle boundary;
- destroy/recreate only required resources;
- reset pair/generated presentation state;
- first new valid pair produces a new midpoint;
- no stale generated output survives the reset.

## G. GPU residency

Normal path must report:

```text
full_frame_cpu_staging == 0
```

Tiny validation-only pixel samples/statistics are allowed.

---

# 11. Timing / telemetry

Record enough data for 01E planning without building a framework.

At minimum:

- valid pair count;
- midpoint submit count;
- midpoint ready count;
- pair ID / generated slot ID;
- flow-to-synthesis dependency wait mechanism;
- midpoint kernel GPU time;
- generated-output D2D/copy time if present;
- D3D11 output map/unmap/submit time where measurable;
- NVOFA-ready -> generated-ready latency;
- input/output-slot reuse rejection count;
- generated stale/pair-mismatch count;
- full-frame CPU staging count;
- per-pair allocation/registration count (must remain zero in normal operation).

Do not optimize solely because a timing number looks improvable in 01D. Record it for 01E/quality-performance work.

---

# 12. Explicitly out of scope

Do **not** implement in NVIDIA-01D unless the minimal path literally cannot function without it:

- learned interpolation/refinement model;
- TensorRT;
- NvOFFRUC;
- forward/softmax splatting;
- hole filling/inpainting;
- depth-aware synthesis;
- sophisticated occlusion masks;
- morphology/dilation stacks;
- UI-specific artifact tuning;
- NVOFA 2x2/1x1 experiments;
- MEDIUM/FAST preset experiments;
- FP16/half2 optimization campaign;
- CUDA Graphs;
- adaptive quality degradation;
- adaptive frame generation;
- sustained deadline contract / broad quality corpus (01E);
- Generic/DirectML/CPU fallback.

A visually imperfect but correctly generated midpoint is acceptable for 01D.

---

# 13. Checkpoint gate

NVIDIA-01D is PASS when this statement is defensible from code and validation evidence:

> The accepted A/B pair and its raw 4x4 NVOFA forward/backward S10.5 flow remain GPU-resident through a native CUDA `t=0.5` midpoint synthesis path; the flow field is reconstructed at full resolution without erroneous vector rescaling, the two endpoint images are backward-warped and fused into a correctly associated full-resolution generated frame, and that generated frame reaches a persistent OBS-visible D3D11 output with explicit pair/resource ownership and no full-frame CPU staging or fallback backend.

Do not block this checkpoint on expected motion-boundary/occlusion artifacts if the transport, math, pair association and generated output are correct.

---

# 14. Git / delivery

Create a dedicated branch from the latest `main`.

Suggested branch:

`feature/nvidia-01d-cuda-midpoint-synthesis`

Deliver NVIDIA-01D as one checkpoint PR.

Before opening the PR:

1. complete the Issue #9 math/source cross-check;
2. inspect the actual SDK 5.0.7 package for any relevant flow-grid upsample/spatial convention before hardcoding it;
3. build the Windows OBS plugin;
4. run static midpoint validation;
5. run known +24 -> +12 translation validation;
6. prove pair/resource ownership across at least two pairs;
7. render a genuine generated midpoint through OBS/D3D11;
8. exercise deterministic real/generated ordering in a controlled smoke;
9. exercise resize/reset lifecycle;
10. record timing/counter evidence;
11. update `PROJECT_STATUS.md` / `TODO.md` only with completed facts.

PR description must include:

- PR number / head SHA;
- primary/peer-reviewed sources inspected;
- exact flow-grid reconstruction convention and evidence;
- S10.5 decode and no-x4 scaling proof;
- exact midpoint-flow math;
- backward-warp coordinate convention;
- border policy;
- source A/B resource used by synthesis;
- flow output resource/stride handling;
- generated buffer/output resource model;
- pair/output ownership lifecycle;
- synchronization chain;
- static test result;
- known-translation midpoint result;
- OBS generated-output proof;
- short real/generated ordering proof;
- resize result;
- midpoint kernel/output timings;
- full-frame CPU staging count;
- known visual limitations;
- confirmation that no forward-splat/learned refinement/01E quality campaign was added;
- confirmation that no fallback/backend cascade was added.

After opening the PR, **stop and wait for PM review. Do not start NVIDIA-01E.**
